File: Sparse Table/14942번 - 개미/14942.cpp
#include <iostream>
#include <vector>

using namespace std;

#define MAX 100001
#define LOG 18
#define X first
#define Y second

int n, energys[MAX];
vector<pair<int, int> > path[MAX];

// i에서 2^j로 이동 후 정점과 필요 에너지
int parents[MAX][LOG];
int dist[MAX][LOG]; 

void input() {
    cin >> n;
    for(int i = 1; i <= n; i++) cin >> energys[i];
    for(int i = 0, a, b, c; i < n - 1; i++) {
        cin >> a >> b >> c;
        path[a].push_back({b, c});
        path[b].push_back({a, c});
    }
}

void dfs(int curNode, int preNode) {
    parents[curNode][0] = preNode;
    
    for(int i = 0; i < path[curNode].size(); i++) {
        int nxtNode = path[curNode][i].X;
        
        if(nxtNode != preNode) {
            dist[nxtNode][0] = path[curNode][i].Y;
            dfs(nxtNode, curNode);
        }
    }
}

void sparse_table() {
    // i섬에서 2^j 이동했을 때의 필요 에너지와 상위 노드 값 테이블 저장
    for(int j = 1; j < LOG; j++) {
        for(int i = 1; i <= n; i++) {
            dist[i][j] = dist[i][j-1] + dist[parents[i][j-1]][j-1];
            parents[i][j] = parents[parents[i][j-1]][j-1];
        }
    }
    
    // 각각의 개미들이 가지고 있는 에너지로 도달할 수 있는 방 출력
    for(int i = 1; i <= n; i++) {
        int pos = i, energy = energys[i];
        for(int j = LOG - 1; j >= 0; j--) {
            if(parents[pos][j] != 0 && energy >= dist[pos][j]) {
                energy -= dist[pos][j];
                pos = parents[pos][j];
            }
        }
        cout << pos << '\n';
    }
}

void solve() {
    dfs(1, 0);
    
    sparse_table();
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Sparse Table/14942번 - 개미/14942.py
import sys

input = sys.stdin.readline

LOG = 20
LIM = 100001

n = int(input())
energys = [0] + [int(input()) for _ in range(n)]
path = [[] for _ in range(n + 1)]
for _ in range(n - 1):
    a, b, c = map(int, input().split())
    path[a].append([b, c])
    path[b].append([a, c])
parents = [ [0] * (LOG) for _ in range(LIM) ]
table = [ [0] * (LOG) for _ in range(LIM) ]
 
def dfs(curNode, preNode):
    parents[curNode][0] = preNode
    
    for nxtNode, nxtDist in path[curNode]:
        if nxtNode != preNode:
            table[nxtNode][0] = nxtDist
            dfs(nxtNode, curNode)
        
def solve():
    dfs(1, 0)
    
    for j in range(1, LOG):
        for i in range(1, n + 1):
            table[i][j] = table[i][j-1] + table[parents[i][j-1]][j-1]
            parents[i][j] = parents[parents[i][j-1]][j-1]
            
    for i in range(1, n + 1):
        pos, energy = i, energys[i]
        for j in reversed(range(LOG)):
            if parents[pos][j] != 0 and energy >= table[pos][j]:
                energy -= table[pos][j]
                pos = parents[pos][j]
        print(pos)


if __name__ == '__main__':
    solve()